### PR TITLE
[processor/filter] fixed vm usage in filterexpr

### PR DESCRIPTION
**Pull request by adam-kiss-sg**
* Added test to filterexpr package to highlight issue with
parallel usage of expression filter.
* Instead of reusing the same vm for all execution of
Matcher.MatchMetric in filterexpr (which is not gorutine-safe), use
a sync.Pool for managing vms.

see: #13573

### CODE REVIEW

**Review by mx-psi — COMMENTED** [review]
Hey there, thanks for the PR, you will need to sign the CLA in order for your changes to be mergeable. I also triggered the test suite so that we can check everything looks good on the CI

**Comment by adam-kiss-sg** [comment]
Thanks @mx-psi. I added the missing changelog entry. Still working on that CLA (company account).

I see that the integration tests failed, but I don't think that's relevant to my changes.

**Comment by adam-kiss-sg** [comment]
@mx-psi 

Hi! My company finally got that CCLA sorted. I did a rebase because the affected files were moved to a different directory. Can you please trigger the tests?

**Comment by mx-psi** [comment]
@frzifus maybe you want to review this? Also cc @boostchicken as the owner

**Comment by TylerHelmuth** [comment]
@adam-kiss-sg @mx-psi there are some pretty significant changes happening to the filterprocessor right now with OTTL that @bogdandrutu and I are working on. I expect at the end of the refactor we won't need expr anymore to interact with datapoint attributes. 

I'm always happy to see more tests and bug fixes and there are still steps we have to take to complete the refactor so I'm happy to see this PR. But know that in the near future expr will probably be deprecated in favor of OTTL.

File: .chloggen/filterexpr-parallel-fix.yaml
Comment (adam-kiss-sg) [reply .chloggen/filterexpr-parallel-fix.yaml]: Makes sense, thanks, updated.

**Comment by adam-kiss-sg** [comment]
@TylerHelmuth thanks for the update. I already switched to the OTTL boolean expression because I needed to filter out datapoints that I couldn't with the expr version (it would only drop or keep whole metrics).

I will leave it up to you whether you want to merge this or not, I'm fine either way :)

**Review by frzifus — COMMENTED** [review]
> @frzifus maybe you want to review this? Also cc @boostchicken as the owner

The change itself, seems valid to me. regarding the ongoing refactoring I am not in the loop.

**Comment by adam-kiss-sg** [comment]
So @TylerHelmuth @mx-psi what's the verdict?

**Review by mx-psi — APPROVED** [review]
I think we should merge this, the component is still out there and we have not officially deprecated it. I will wait for @TylerHelmuth to review before merging

**Review by TylerHelmuth — APPROVED** [review]
@mx-psi yes thats fine, those config options are explicitly NOT deprecated yet bc there are some feature gaps to fill in still (https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16413) and this PR does not add any more gaps.

**Pull request merged**
